src/sql/src/plan/statement/ddl.rs

```diff
@@ -372,39 +371,35 @@ pub fn plan_create_source(
 
     let (external_connection, encoding) = match connection {
         CreateSourceConnection::Kafka(kafka) => {
-            let mut options = kafka_util::extract_config(&mut legacy_with_options)?;
-            let kafka_connection = match &kafka.connection {
+            let (kafka_connection, options) = match &kafka.connection {
                 mz_sql_parser::ast::KafkaConnection::Inline { broker } => {
                     scx.require_unsafe_mode("creating Kafka sources with inline connections")?;
+                    let mut options = BTreeMap::new();
                     options.insert(
                         "bootstrap.servers".into(),
                         KafkaAddrs::from_str(broker)
                             .map_err(|e| sql_err!("parsing kafka broker: {e}"))?
                             .to_string()
                             .into(),
                     );
-                    KafkaConnection::try_from(&mut options)?
+                    let connection = KafkaConnection::try_from(&mut options)?;
+                    (connection, options)
                 }
-                mz_sql_parser::ast::KafkaConnection::Reference { connection, .. } => {
+                mz_sql_parser::ast::KafkaConnection::Reference {
+                    connection,
+                    with_options,
+                } => {
                     let item = scx.get_item_by_resolved_name(&connection)?;
                     let connection = match item.connection()? {
                         Connection::Kafka(connection) => connection.clone(),
                         _ => sql_bail!("{} is not a kafka connection", item.name()),
                     };
 
-                    // TODO: Once we remove use of `String`-keyed options, we
-                    // can remove this check because permitted values here will
-                    // simply be a disjoint set of what `CONNECTION`s support.
-                    for k in BTreeMap::<String, StringOrSecret>::from(connection.clone()).keys() {
-                        if options.contains_key(k) {
-                            sql_bail!(
-                                "cannot set option {} for SOURCE using CONNECTION {}",
-                                k,
-                                scx.catalog.resolve_full_name(item.name())
-                            );
-                        }
-                    }
-                    connection
+                    let with_options: KafkaConfigOptionExtracted =
+                        with_options.clone().try_into()?;
+                    let with_options: BTreeMap<String, StringOrSecret> = with_options.try_into()?;
```

Comment:
I have a hunch that we don't actually want to allow anyone to configure these options in platform. They currently require unsafe mode to use (because _all_ legacy WITH options require unsafe mode to use), and I'd love to preserve that. Can you add a check here and below in the create sink path that requires unsafe mode if `with_options` is not empty?